SQL failed with "ERROR 1105 (HY000): runtime error: invalid memory address or nil pointer dereference" when tidb_enable_inl_join_inner_multi_pattern is ON

## Bug Report

### 1. Minimal reproduce step (Required)

```sql
set GLOBAL tidb_enable_inl_join_inner_multi_pattern='ON';
create table t1(col_1 int, index idx_1(col_1));
create table t2(col_1 int, col_2 int, index idx_2(col_1));

select /*+ inl_join(tmp) */ * from t1 inner join (select col_1, group_concat(col_2) from t2 group by col_1) tmp on t1.col_1 = tmp.col_1;
```

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

Run successfully

### 3. What did you see instead (Required)

```
[2024/08/02 15:51:11.915 +08:00] [ERROR] [conn.go:1040] ["connection running loop panic"] [conn=2097156] [session_alias=] [lastSQL="select /*+ inl_join(tmp) */ * from t1 inner join (select col_1, group_concat(col_2) from t2 group by col_1) tmp on t1.col_1 = tmp.col_1"] [err="runtime error: invalid memory address or nil pointer dereference"] [stack="github.com/pingcap/tidb/pkg/server.(*clientConn).Run.func1\n\t/Users/joechenrh/code/tidb-fix/pkg/server/conn.go:1043\nruntime.gopanic
/Users/joechenrh/deps/go/src/runtime/panic.go:770\ngithub.com/pingcap/tidb/pkg/executor.(*Compiler).Compile.func1
/Users/joechenrh/code/tidb-fix/pkg/executor/compiler.go:57\nruntime.gopanic
/Users/joechenrh/deps/go/src/runtime/panic.go:770\nruntime.panicmem
/Users/joechenrh/deps/go/src/runtime/panic.go:261\nruntime.sigpanic
/Users/joechenrh/deps/go/src/runtime/signal_unix.go:881\ngithub.com/pingcap/tidb/pkg/planner/property.(*PhysicalProperty).CloneEssentialFields
/Users/joechenrh/code/tidb-fix/pkg/planner/property/physical_property.go:366\ngithub.com/pingcap/tidb/pkg/planner/core.InjectProjBelowAgg
/Users/joechenrh/code/tidb-fix/pkg/planner/core/rule_inject_extra_projection.go:208\ngithub.com/pingcap/tidb/pkg/planner/core.(*projInjector).inject
/Users/joechenrh/code/tidb-fix/pkg/planner/core/rule_inject_extra_projection.go:68\ngithub.com/pingcap/tidb/pkg/planner/core.(*projInjector).inject\n\t/Users/joechenrh/code/tidb-fix/pkg/planner/core/rule_inject_extra_projection.go:58\ngithub.com/pingcap/tidb/pkg/planner/core.(*projInjector).inject
/Users/joechenrh/code/tidb-fix/pkg/planner/core/rule_inject_extra_projection.go:58\ngithub.com/pingcap/tidb/pkg/planner/core.InjectExtraProjection
/Users/joechenrh/code/tidb-fix/pkg/planner/core/rule_inject_extra_projection.go:45\ngithub.com/pingcap/tidb/pkg/planner/core.postOptimize
/Users/joechenrh/code/tidb-fix/pkg/planner/core/optimizer.go:424\ngithub.com/pingcap/tidb/pkg/planner/core.doOptimize
/Users/joechenrh/code/tidb-fix/pkg/planner/core/optimizer.go:312\ngithub.com/pingcap/tidb/pkg/planner/core.DoOptimize
/Users/joechenrh/code/tidb-fix/pkg/planner/core/optimizer.go:355\ngithub.com/pingcap/tidb/pkg/planner.optimize
/Users/joechenrh/code/tidb-fix/pkg/planner/optimize.go:525\ngithub.com/pingcap/tidb/pkg/planner.Optimize
/Users/joechenrh/code/tidb-fix/pkg/planner/optimize.go:356\ngithub.com/pingcap/tidb/pkg/executor.(*Compiler).Compile
/Users/joechenrh/code/tidb-fix/pkg/executor/compiler.go:99\ngithub.com/pingcap/tidb/pkg/session.(*session).ExecuteStmt
/Users/joechenrh/code/tidb-fix/pkg/session/session.go:2097\ngithub.com/pingcap/tidb/pkg/server.(*TiDBContext).ExecuteStmt
/Users/joechenrh/code/tidb-fix/pkg/server/driver_tidb.go:291\ngithub.com/pingcap/tidb/pkg/server.(*clientConn).handleStmt
/Users/joechenrh/code/tidb-fix/pkg/server/conn.go:2047\ngithub.com/pingcap/tidb/pkg/server.(*clientConn).handleQuery
/Users/joechenrh/code/tidb-fix/pkg/server/conn.go:1801\ngithub.com/pingcap/tidb/pkg/server.(*clientConn).dispatch
/Users/joechenrh/code/tidb-fix/pkg/server/conn.go:1375\ngithub.com/pingcap/tidb/pkg/server.(*clientConn).Run
/Users/joechenrh/code/tidb-fix/pkg/server/conn.go:1141\ngithub.com/pingcap/tidb/pkg/server.(*Server).onConn
/Users/joechenrh/code/tidb-fix/pkg/server/server.go:740"]
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

--- a/pkg/planner/core/exhaust_physical_plans.go
+++ b/pkg/planner/core/exhaust_physical_plans.go
@@ -1026,44 +1026,44 @@ func constructInnerTableScanTask(
 	return constructIndexJoinInnerSideTask(p, prop, copTask, ds, nil, wrapper)
 }
 
-func constructInnerByZippedChildren(zippedChildren []base.LogicalPlan, child base.PhysicalPlan) base.PhysicalPlan {
+func constructInnerByZippedChildren(prop *property.PhysicalProperty, zippedChildren []base.LogicalPlan, child base.PhysicalPlan) base.PhysicalPlan {
 	for i := len(zippedChildren) - 1; i >= 0; i-- {
 		switch x := zippedChildren[i].(type) {
 		case *LogicalUnionScan:
-			child = constructInnerUnionScan(x, child)
+			child = constructInnerUnionScan(prop, x, child)
 		case *logicalop.LogicalProjection:
-			child = constructInnerProj(x, child)
+			child = constructInnerProj(prop, x, child)
 		case *LogicalSelection:
-			child = constructInnerSel(x, child)
+			child = constructInnerSel(prop, x, child)
 		case *LogicalAggregation:
-			child = constructInnerAgg(x, child)
+			child = constructInnerAgg(prop, x, child)
 		}
 	}
 	return child
 }
 
-func constructInnerAgg(logicalAgg *LogicalAggregation, child base.PhysicalPlan) base.PhysicalPlan {
+func constructInnerAgg(prop *property.PhysicalProperty, logicalAgg *LogicalAggregation, child base.PhysicalPlan) base.PhysicalPlan {
 	if logicalAgg == nil {
 		return child
 	}
-	physicalHashAgg := NewPhysicalHashAgg(logicalAgg, logicalAgg.StatsInfo(), nil)
+	physicalHashAgg := NewPhysicalHashAgg(logicalAgg, logicalAgg.StatsInfo(), prop)
 	physicalHashAgg.SetSchema(logicalAgg.Schema().Clone())
 	physicalHashAgg.SetChildren(child)
 	return physicalHashAgg
 }
 
-func constructInnerSel(sel *LogicalSelection, child base.PhysicalPlan) base.PhysicalPlan {
+func constructInnerSel(prop *property.PhysicalProperty, sel *LogicalSelection, child base.PhysicalPlan) base.PhysicalPlan {
 	if sel == nil {
 		return child
 	}
 	physicalSel := PhysicalSelection{
 		Conditions: sel.Conditions,
-	}.Init(sel.SCtx(), sel.StatsInfo(), sel.QueryBlockOffset(), nil)
+	}.Init(sel.SCtx(), sel.StatsInfo(), sel.QueryBlockOffset(), prop)
 	physicalSel.SetChildren(child)
 	return physicalSel
 }
 
-func constructInnerProj(proj *logicalop.LogicalProjection, child base.PhysicalPlan) base.PhysicalPlan {
+func constructInnerProj(prop *property.PhysicalProperty, proj *logicalop.LogicalProjection, child base.PhysicalPlan) base.PhysicalPlan {
 	if proj == nil {
 		return child
 	}
@@ -1071,13 +1071,13 @@ func constructInnerProj(proj *logicalop.LogicalProjection, child base.PhysicalPl
 		Exprs:                proj.Exprs,
 		CalculateNoDelay:     proj.CalculateNoDelay,
 		AvoidColumnEvaluator: proj.AvoidColumnEvaluator,
-	}.Init(proj.SCtx(), proj.StatsInfo(), proj.QueryBlockOffset(), nil)
+	}.Init(proj.SCtx(), proj.StatsInfo(), proj.QueryBlockOffset(), prop)
 	physicalProj.SetChildren(child)
 	physicalProj.SetSchema(proj.Schema())
 	return physicalProj
 }
 
-func constructInnerUnionScan(us *LogicalUnionScan, reader base.PhysicalPlan) base.PhysicalPlan {
+func constructInnerUnionScan(prop *property.PhysicalProperty, us *LogicalUnionScan, reader base.PhysicalPlan) base.PhysicalPlan {
 	if us == nil {
 		return reader
 	}
@@ -1086,7 +1086,7 @@ func constructInnerUnionScan(us *LogicalUnionScan, reader base.PhysicalPlan) bas
 	physicalUnionScan := PhysicalUnionScan{
 		Conditions: us.Conditions,
 		HandleCols: us.HandleCols,
-	}.Init(us.SCtx(), reader.StatsInfo(), us.QueryBlockOffset(), nil)
+	}.Init(us.SCtx(), reader.StatsInfo(), us.QueryBlockOffset(), prop)
 	physicalUnionScan.SetChildren(reader)
 	return physicalUnionScan
 }
@@ -1353,7 +1353,7 @@ func constructIndexJoinInnerSideTask(p *LogicalJoin, prop *property.PhysicalProp
 	// If the bottom plan is not aggregation or the aggregation can't be pushed to coprocessor, we will construct a root task directly.
 	if !canPushAggToCop {
 		result := dsCopTask.ConvertToRootTask(ds.SCtx()).(*RootTask)
-		result.SetPlan(constructInnerByZippedChildren(wrapper.zippedChildren, result.GetPlan()))
+		result.SetPlan(constructInnerByZippedChildren(prop, wrapper.zippedChildren, result.GetPlan()))
 		return result
 	}
 
@@ -1434,7 +1434,7 @@ func constructIndexJoinInnerSideTask(p *LogicalJoin, prop *property.PhysicalProp
 	if !ok {
 		return nil
 	}
-	result.SetPlan(constructInnerByZippedChildren(wrapper.zippedChildren[0:len(wrapper.zippedChildren)-1], result.p))
+	result.SetPlan(constructInnerByZippedChildren(prop, wrapper.zippedChildren[0:len(wrapper.zippedChildren)-1], result.p))
 	return result
 }
 

--- a/pkg/planner/core/exhaust_physical_plans.go
+++ b/pkg/planner/core/exhaust_physical_plans.go
@@ -814,13 +814,13 @@ func buildIndexJoinInner2TableScan(
 			return nil
 		}
 		rangeInfo := indexJoinPathRangeInfo(p.SCtx(), outerJoinKeys, indexJoinResult)
-		innerTask = constructInnerTableScanTask(p, wrapper, indexJoinResult.chosenRanges.Range(), outerJoinKeys, rangeInfo, false, false, avgInnerRowCnt)
+		innerTask = constructInnerTableScanTask(p, prop, wrapper, indexJoinResult.chosenRanges.Range(), outerJoinKeys, rangeInfo, false, false, avgInnerRowCnt)
 		// The index merge join's inner plan is different from index join, so we
 		// should construct another inner plan for it.
 		// Because we can't keep order for union scan, if there is a union scan in inner task,
 		// we can't construct index merge join.
 		if !wrapper.hasDitryWrite {
-			innerTask2 = constructInnerTableScanTask(p, wrapper, indexJoinResult.chosenRanges.Range(), outerJoinKeys, rangeInfo, true, !prop.IsSortItemEmpty() && prop.SortItems[0].Desc, avgInnerRowCnt)
+			innerTask2 = constructInnerTableScanTask(p, prop, wrapper, indexJoinResult.chosenRanges.Range(), outerJoinKeys, rangeInfo, true, !prop.IsSortItemEmpty() && prop.SortItems[0].Desc, avgInnerRowCnt)
 		}
 		ranges = indexJoinResult.chosenRanges
 	} else {
@@ -854,13 +854,13 @@ func buildIndexJoinInner2TableScan(
 		}
 		buffer.WriteString("]")
 		rangeInfo := buffer.String()
-		innerTask = constructInnerTableScanTask(p, wrapper, ranges, outerJoinKeys, rangeInfo, false, false, avgInnerRowCnt)
+		innerTask = constructInnerTableScanTask(p, prop, wrapper, ranges, outerJoinKeys, rangeInfo, false, false, avgInnerRowCnt)
 		// The index merge join's inner plan is different from index join, so we
 		// should construct another inner plan for it.
 		// Because we can't keep order for union scan, if there is a union scan in inner task,
 		// we can't construct index merge join.
 		if !wrapper.hasDitryWrite {
-			innerTask2 = constructInnerTableScanTask(p, wrapper, ranges, outerJoinKeys, rangeInfo, true, !prop.IsSortItemEmpty() && prop.SortItems[0].Desc, avgInnerRowCnt)
+			innerTask2 = constructInnerTableScanTask(p, prop, wrapper, ranges, outerJoinKeys, rangeInfo, true, !prop.IsSortItemEmpty() && prop.SortItems[0].Desc, avgInnerRowCnt)
 		}
 	}
 	var (
@@ -922,7 +922,7 @@ func buildIndexJoinInner2IndexScan(
 			maxOneRow = ok && (sf.FuncName.L == ast.EQ)
 		}
 	}
-	innerTask := constructInnerIndexScanTask(p, wrapper, indexJoinResult.chosenPath, indexJoinResult.chosenRanges.Range(), indexJoinResult.chosenRemained, innerJoinKeys, indexJoinResult.idxOff2KeyOff, rangeInfo, false, false, avgInnerRowCnt, maxOneRow)
+	innerTask := constructInnerIndexScanTask(p, prop, wrapper, indexJoinResult.chosenPath, indexJoinResult.chosenRanges.Range(), indexJoinResult.chosenRemained, innerJoinKeys, indexJoinResult.idxOff2KeyOff, rangeInfo, false, false, avgInnerRowCnt, maxOneRow)
 	failpoint.Inject("MockOnlyEnableIndexHashJoin", func(val failpoint.Value) {
 		if val.(bool) && !p.SCtx().GetSessionVars().InRestrictedSQL && innerTask != nil {
 			failpoint.Return(constructIndexHashJoin(p, prop, outerIdx, innerTask, indexJoinResult.chosenRanges, keyOff2IdxOff, indexJoinResult.chosenPath, indexJoinResult.lastColManager))
@@ -939,7 +939,7 @@ func buildIndexJoinInner2IndexScan(
 	// Because we can't keep order for union scan, if there is a union scan in inner task,
 	// we can't construct index merge join.
 	if !wrapper.hasDitryWrite {
-		innerTask2 := constructInnerIndexScanTask(p, wrapper, indexJoinResult.chosenPath, indexJoinResult.chosenRanges.Range(), indexJoinResult.chosenRemained, innerJoinKeys, indexJoinResult.idxOff2KeyOff, rangeInfo, true, !prop.IsSortItemEmpty() && prop.SortItems[0].Desc, avgInnerRowCnt, maxOneRow)
+		innerTask2 := constructInnerIndexScanTask(p, prop, wrapper, indexJoinResult.chosenPath, indexJoinResult.chosenRanges.Range(), indexJoinResult.chosenRemained, innerJoinKeys, indexJoinResult.idxOff2KeyOff, rangeInfo, true, !prop.IsSortItemEmpty() && prop.SortItems[0].Desc, avgInnerRowCnt, maxOneRow)
 		if innerTask2 != nil {
 			joins = append(joins, constructIndexMergeJoin(p, prop, outerIdx, innerTask2, indexJoinResult.chosenRanges, keyOff2IdxOff, indexJoinResult.chosenPath, indexJoinResult.lastColManager)...)
 		}
@@ -950,6 +950,7 @@ func buildIndexJoinInner2IndexScan(
 // constructInnerTableScanTask is specially used to construct the inner plan for PhysicalIndexJoin.
 func constructInnerTableScanTask(
 	p *LogicalJoin,
+	prop *property.PhysicalProperty,
 	wrapper *indexJoinInnerChildWrapper,
 	ranges ranger.Ranges,
 	_ []*expression.Column,
@@ -1022,7 +1023,7 @@ func constructInnerTableScanTask(
 	ts.PlanPartInfo = copTask.physPlanPartInfo
 	selStats := ts.StatsInfo().Scale(selectivity)
 	ts.addPushedDownSelection(copTask, selStats)
-	return constructIndexJoinInnerSideTask(p, copTask, ds, nil, wrapper)
+	return constructIndexJoinInnerSideTask(p, prop, copTask, ds, nil, wrapper)
 }
 
 func constructInnerByZippedChildren(zippedChildren []base.LogicalPlan, child base.PhysicalPlan) base.PhysicalPlan {
@@ -1142,6 +1143,7 @@ func getColsNDVLowerBoundFromHistColl(colUIDs []int64, histColl *statistics.Hist
 // constructInnerIndexScanTask is specially used to construct the inner plan for PhysicalIndexJoin.
 func constructInnerIndexScanTask(
 	p *LogicalJoin,
+	prop *property.PhysicalProperty,
 	wrapper *indexJoinInnerChildWrapper,
 	path *util.AccessPath,
 	ranges ranger.Ranges,
@@ -1323,7 +1325,7 @@ func constructInnerIndexScanTask(
 		logutil.BgLogger().Warn("unexpected error happened during addPushedDownSelection function", zap.Error(err))
 		return nil
 	}
-	return constructIndexJoinInnerSideTask(p, cop, ds, path, wrapper)
+	return constructIndexJoinInnerSideTask(p, prop, cop, ds, path, wrapper)
 }
 
 // construct the inner join task by inner child plan tree
@@ -1334,7 +1336,7 @@ func constructInnerIndexScanTask(
 //	There are two kinds of agg: stream agg and hash agg. Stream agg depends on some conditions, such as the group by cols
 //
 // Step2: build other inner plan node to task
-func constructIndexJoinInnerSideTask(p *LogicalJoin, dsCopTask *CopTask, ds *DataSource, path *util.AccessPath, wrapper *indexJoinInnerChildWrapper) base.Task {
+func constructIndexJoinInnerSideTask(p *LogicalJoin, prop *property.PhysicalProperty, dsCopTask *CopTask, ds *DataSource, path *util.AccessPath, wrapper *indexJoinInnerChildWrapper) base.Task {
 	var la *LogicalAggregation
 	var canPushAggToCop bool
 	if len(wrapper.zippedChildren) > 0 {
@@ -1403,7 +1405,7 @@ func constructIndexJoinInnerSideTask(p *LogicalJoin, dsCopTask *CopTask, ds *Dat
 		streamAgg := basePhysicalAgg{
 			GroupByItems: newGbyItems,
 			AggFuncs:     newAggFuncs,
-		}.initForStream(la.SCtx(), la.StatsInfo(), la.QueryBlockOffset(), nil)
+		}.initForStream(la.SCtx(), la.StatsInfo(), la.QueryBlockOffset(), prop)
 		streamAgg.SetSchema(la.Schema().Clone())
 		// change to keep order for index scan and dsCopTask
 		if dsCopTask.indexPlan != nil {
@@ -1422,7 +1424,7 @@ func constructIndexJoinInnerSideTask(p *LogicalJoin, dsCopTask *CopTask, ds *Dat
 
 	// build hash agg, when the stream agg is illegal such as the order by prop is not matched
 	if aggTask == nil {
-		physicalHashAgg := NewPhysicalHashAgg(la, la.StatsInfo(), nil)
+		physicalHashAgg := NewPhysicalHashAgg(la, la.StatsInfo(), prop)
 		physicalHashAgg.SetSchema(la.Schema().Clone())
 		aggTask = physicalHashAgg.Attach2Task(dsCopTask)
 	}

--- a/pkg/planner/core/issuetest/planner_issue_test.go
+++ b/pkg/planner/core/issuetest/planner_issue_test.go
@@ -112,3 +112,13 @@ func TestIssue54535(t *testing.T) {
 			"      └─HashAgg_13(Probe) 79840080.00 cop[tikv]  group by:test.tb.b1, test.tb.b2, funcs:count(test.tb.b3)->Column#11",
 			"        └─TableRowIDScan_11 9990.00 cop[tikv] table:tb keep order:false, stats:pseudo"))
 }
+
+func Test55169(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("set session tidb_enable_inl_join_inner_multi_pattern='ON';")
+	tk.MustExec("create table t1(col_1 int, index idx_1(col_1));")
+	tk.MustExec("create table t2(col_1 int, col_2 int, index idx_2(col_1));")
+	tk.MustExec("select /*+ inl_join(tmp) */ * from t1 inner join (select col_1, group_concat(col_2) from t2 group by col_1) tmp on t1.col_1 = tmp.col_1;")
+}

--- a/pkg/planner/core/issuetest/planner_issue_test.go
+++ b/pkg/planner/core/issuetest/planner_issue_test.go
@@ -115,4 +115,5 @@ func TestIssue54535(t *testing.T) {
 	tk.MustExec("create table t1(col_1 int, index idx_1(col_1));")
 	tk.MustExec("create table t2(col_1 int, col_2 int, index idx_2(col_1));")
 	tk.MustQuery("select /*+ inl_join(tmp) */ * from t1 inner join (select col_1, group_concat(col_2) from t2 group by col_1) tmp on t1.col_1 = tmp.col_1;").Check(testkit.Rows())
+	tk.MustQuery("select /*+ inl_join(tmp) */ * from t1 inner join (select col_1, group_concat(distinct col_2 order by col_2) from t2 group by col_1) tmp on t1.col_1 = tmp.col_1;").Check(testkit.Rows())
 }

--- a/pkg/planner/core/issuetest/planner_issue_test.go
+++ b/pkg/planner/core/issuetest/planner_issue_test.go
@@ -111,14 +111,8 @@ func TestIssue54535(t *testing.T) {
 			"      │ └─IndexRangeScan_10 10000.00 cop[tikv] table:tb, index:idx_b(b1) range: decided by [eq(test.tb.b1, test.ta.a1)], keep order:false, stats:pseudo",
 			"      └─HashAgg_13(Probe) 79840080.00 cop[tikv]  group by:test.tb.b1, test.tb.b2, funcs:count(test.tb.b3)->Column#11",
 			"        └─TableRowIDScan_11 9990.00 cop[tikv] table:tb keep order:false, stats:pseudo"))
-}
-
-func Test55169(t *testing.T) {
-	store := testkit.CreateMockStore(t)
-	tk := testkit.NewTestKit(t, store)
-	tk.MustExec("use test")
-	tk.MustExec("set session tidb_enable_inl_join_inner_multi_pattern='ON';")
+	// test for issues/55169
 	tk.MustExec("create table t1(col_1 int, index idx_1(col_1));")
 	tk.MustExec("create table t2(col_1 int, col_2 int, index idx_2(col_1));")
-	tk.MustExec("select /*+ inl_join(tmp) */ * from t1 inner join (select col_1, group_concat(col_2) from t2 group by col_1) tmp on t1.col_1 = tmp.col_1;")
+	tk.MustQuery("select /*+ inl_join(tmp) */ * from t1 inner join (select col_1, group_concat(col_2) from t2 group by col_1) tmp on t1.col_1 = tmp.col_1;").Check(testkit.Rows())
 }

